### PR TITLE
New Feedback `init` driven by external control signal

### PR DESCRIPTION
**Pull request by p4checo**
# Motivation 

When state transitions are to be triggered from external events (e.g. reacting to user interaction, notification, etc) via `Signal`'s, all existing `init`s except the "raw" `init(effects:)` use the system's `state` signal as the "driver" of the `events` chain.

When used with the previous `init`s, these external events can be processed "out of sync" by the system from a time perspective, because when the `query` or `predicate` "passes" (i.e. it returns a non `nil` `Control` or `true`) but the inner `Effect` still hasn't produced a value, the chain will block until a value is returned by the external signal.

On some cases, the `Effect` might not even make sense anymore, because by that time other state transitions might already have occurred, and the `query`/`predicate` of the Feedback wouldn't "pass" under the current state anymore. Since they have already passed on the chain, the `Effect` is processed anyway, which will result in essentially an invalid event (and possibly transition).

For these scenarios, it appears that the external signal should be the "driver" of the `events` chain, so that whenever it fires a control value, the current system's state is captured, and both values are then
evaluated using a closure (similar to `query`) that produces a new effect (`Event`), or cancels any previous one.

To address this, a new `init(control:effect:)` was created on `Feedback` to provide this behavior.

## Notes

Haven't added any tests or additional documentation because I would like gather some input from you guys first, to know if these changes make sense at all, and don't go against the Feedback "concept" 😉

### CODE REVIEW

**Comment by andersio** [comment]
In such model, when a control-plus-state derived event hits the reducer, the state could have potentially been moved away from the earlier one that derives the event. This causes a legit consistency concern, because now events could be sourced from two desynchronised timeline of state in a concurrent environment.

IMO it is preferable to include _all_ externally triggered events as part of the system `Event` set, and have them centrally processed by the reducer. The primitive `Feedback.init` accepting two-argument setup closures can be used for such purpose — for example:

```swift
enum UserAction { /* Representing all UI interactions */ }
enum Event {
    /* Some internal events */

    // Include all user interaction events.
    case ui(UserAction)
}

let (userActions, userActionObserver) = Signal<UserAction, NoError>.pipe()
let feedback = Feedback { scheduler, _ in
    return userActions.map(Event.ui).observe(on: scheduler)
}
```

**Comment by p4checo** [comment]
> In such model, when the derived event hits the reducer, the state could have potentially been updated from the earlier one that derives the event. This causes a legit consistency concern, because now events could be sourced from two desynchronised timeline of state in a concurrent environment.

Not sure if I followed this, sorry. 😓 Is your concern about the state having potentially been updated due to the fact that we're using `withLatest` to capture the `state`?
Because what you describe seems almost what this `init` is trying to solve, for some scenarios.

> IMO it is preferable to include all externally triggered events as part of the system Event set, and have them centrally processed by the reducer. The primitive Feedback.init accepting _two-argument setup closures can be used for such purpose

Your example definitely can be used on some scenarios, but it doesn't support acting on the current state whenever the external signal fires. In some use cases, you might want to discard these same external values according to the state of the system, or simply use the captured state to decide which event to emit. For instance, on our app we use `Feedback`s to manage lazy loading, and to trigger a new batch request we now have the following (using the new `init`):

```swift
static func loadNextBatchFeedback(for nearBottom: Signal<Void, NoError>) -> Feedback<State, Event> {
    return Feedback(control: nearBottom, 
                    effect: { _, state in state.canLoadNextBatch ? Event.loadNextBatch : nil })
}
```

This ensures that whenever the external signal fires, we will trigger an `Event.loadNextBatch` event as long as the captured state allows it. Since it runs on the system's scheduler, I am assuming that it doesn't introduce any consistency problem due to concurrency. I might be completely wrong though 😅

Previously, we had this (based on your example):

```swift
static func loadNextBatchFeedback(for nearBottom: Signal<Void, NoError>) -> Feedback<State, Event> {
    return Feedback(predicate: { $0.canLoadNextBatch },
                    effects: { _  in nearBottom.map { Event.loadNextBatch } })
}
```

However, we had issues using this primitive, where the `Event.loadNextBatch` would be emitted out of sync, and more times than it should. Does it make sense? The observed scenario was precisely the one you described above, where events came from two desynchronised timelines of state.

One important thing to note too, is that we're using `UIScheduler` for the `system`, which may have influence on this issue manifesting itself.

**Comment by sergdort** [comment]
Hi, guys! 

It seems to me that the way we used to build our apps is becoming a source of confusion in designing the `Feedback` systems (and probably we should improve our documentation somehow). I mean that we tend to design our `ViewModels` (or whatever we prefer to use as a sync point between Business logic and UI)  where user input considered as an action to perform some side effects, but with `Feedback` -based approach, the `State` in a way acts as a source of truth  as well as it may act as a control unit to decide whether to perform any `Effects` based on the current value of the state. This is where  `init<Control: Equatable>(control: (State) -> Control?, effects: (Control) -> Effect)` overload can be handy.

So back to the example with pagination:
User scrolling to the end of the view is a "suggestion" to perform effects e.g we emit `Event.loadNext` and then `Reducer`  transitioning to the `State` where the value of the `state.nextPage` changes which should perform loading of the next page. This means that we should design our `Reducer` or loadNextPage `Feedback` in such way to be able to prevent any changes to the `state.nextPage` if several event where emitted from the `scrollView` (Because it will cancel performed side effects)

1. Example with reducer would be to ignore `Event.loadNext` if we are in the load next page`State`
2. Example with the `Feedback` would be something like: 
```swift
func loadNextPage(trigger: Signal<Void, NoError>) -> Feedback<State, Event> {
    return Feedback { state in
        guard state = .loadingNextPage else {
                return .empty
        }
        return trigger.map(Event.loadNext)
    }
}
```

**Comment by p4checo** [comment]
Hi 👋 

I still struggle to fully grasp the entire Feedback "philosophy" (and inner mechanics), but I can already see it's huge potential in managing state. 😉 

I also thought about simply filtering these "situations" in the `reducer`, but unfortunately it always has to return a `Value` (state), so we can't exactly discard any `Event` that comes "out of sync", "duplicated" and/or would result in an invalid state transition.

Perhaps if the `reduce` allowed `nil`, i.e. `@escaping (Value, Event) -> Value?`, then we could easily switch on both `Value` and `Event` and only allow valid combinations to trigger a state change. Once again, not sure if this makes sense or goes against the Feedback / Reducer "concept". :innocent: 

From my understanding, all `Event`'s that are emitted should already have been "validated" by either a `query` or a `predicate`, so that they result in a valid state transition. The provided `init`'s work fine for "internal" events (`state` triggered), but appear to have some limitations/issues when used in conjunction with an "external" source (from a `Signal` that might not have produced any value when the state changes). The only solution I found so far was to invert their "priorities" in the `events` signal.

Meanwhile, I've made an experiment of using a `UIScheduler` as the `Property`'s `scheduler` in the pagination screen of the example app (`PaginationViewController.swift:94`), and it stopped working (no cells are shown). I guess using this scheduler does indeed create some entropy after all 😆

Another thing I don't understand from the example is why it's doing `.observe(on: UIScheduler())` on both `pagingFeedback()` and `retryPagingFeedback()`, given the effects are already observed using `.observe(on: scheduler)` (which is using the default `QueueScheduler.main` already). 🤔

**Comment by sergdort** [comment]
> Another thing I don't understand from the example is why it's doing .observe(on: UIScheduler())

Yeah it's absolutely unnecessary :)

**Comment by andersio** [comment]
> Not sure if I followed this, sorry. 😓 Is your concern about the state having potentially been updated due to the fact that we're using withLatest to capture the state?
Because what you describe seems almost what this init is trying to solve, for some scenarios.

I assume it should be fine for UI binding purpose. My only concern is that when people use this variant of feedback to feed values combined with the captured state back into the loop, effectively bypassing the reducer.

Here is a simple picture:

1. I got two 🐈 and one unoccupied 🛋 in my place.
2. An external trigger `rescuedCats` takes the current state and make the decision to put the rescued 🐈 on the 🛋. This leads to an event `update([🐈, 🐈, 🐈], sofa: .occupied)` to be scheduled on the feedback loop's home scheduler.
3. When the scheduler ticks, the event gets evaluated by the reducer, and the new state of three 🐈 and one occupied 🛋 is produced.

This looks fine. But now consider that the feedback loop itself also models a unblocked backyard entrance for stray 🐈, which can come and go at any time. If any update due to this happens between (2) and (3), the end state gonna be incorrect.

We could argue that this is due to the state machine being incorrectly modelled, just that my worry is that such variant of initialiser encourages such pattern, or at least makes it easier to be done.

> I also thought about simply filtering these "situations" in the reducer, but unfortunately it always has to return a Value (state), so we can't exactly discard any Event that comes "out of sync", "duplicated" and/or would result in an invalid state transition.

IMO it is fine to look at both the current state and the event together to determine the next state.

It is pretty common for us to return the same state when unanticipated and/or repeated events happen. We also deal with the fact of repeated state by utilising repeat-skipping feedbacks (e.g. skip repeated based on `Equatable`, or "first until nil", whichever appropriate).

**Comment by p4checo** [comment]
Thanks for the cool explanation! 😹🙏 

> This looks fine. But now consider that the feedback loop itself also models a unblocked backyard entrance for stray 🐈, which can come and go at any time. If any update due to this happens between (2) and (3), the end state gonna be incorrect.

If I understood this correctly, the problem is that only when the event `update([🐈, 🐈, 🐈], sofa: .occupied)` is sent is it scheduled on the `system`s scheduler (i.e. via `observe(on:)`). Because of this, a time frame exists where the new event isn't thread safe with any other events that might occur on the state machine, possibly leading to inconsistencies.

Following this reasoning, the problem can then be narrowed down to the scheduling on the system's `scheduler` "being too late" in the chain ,right? If the event was scheduled on the `scheduler` right after being emitted by the external source, and only afterwards the state was captured and the effect (event) calculated, it would no longer be possible for other events to be fired in the meantime, since they would have to wait on the scheduler's queue (and please correct me if I'm wrong 😛).

What if we change the chain a bit to this:

```swift
public init<Control>(control: Signal<Control, NoError>, effect: @escaping (Control, State) -> Event?) {
    self.events = { scheduler, state in
        control.
            .observe(on: scheduler)
            .withLatest(from: state)
            .flatMap(.latest) { control, state -> SignalProducer<Event, NoError> in
                guard let event = effect(control, state) else { return .empty }
                return SignalProducer(value: event)
            }
    }
}
```

Do you think we could address the problem you are mentioning?

The main downside I see is that it can "spam" the system's scheduler with values that can be discarded by the `effect` closure, but it seems to me that we have more guarantees about the state's consistency.

> IMO it is fine to look at both the current state and the event together to determine the next state.

Definitely! It allows for much more complex state machines. My point was simply that the reducer doesn't allow one to skip an event without producing a state.

> It is pretty common for us to return the same state when unanticipated and/or repeated events happen. 

Do you perform this on the final state of the system (i.e. reducer's output)? If so, honestly, (even though it works) I see resorting to repeat skipping states on a state machine to be a bit "smelly", unless the state machine explicitly requires "cyclic" states (which I think most of the times is not the case?).  This is precisely why I was suggesting a reducer that could return an optional State (i.e. to allow modelling invalid transitions/states as`nil`, and ignoring them with `filterMap`).

> We also deal with the fact of repeated state by utilising repeat-skipping feedbacks (e.g. skip repeated based on Equatable, or "first until nil", whichever appropriate).

On the other hand, performing this on the feedbacks (which are essentially the "state transition generators") seems perfectly fine (and correct), since you are simply controlling the events that are generated (so no invalid/inconsistent states are derived).

**Comment by andersio** [comment]
For this to work ideally, IMO we would need a custom operator that combines `flatMap(.latest)` and `withLatest(from:)` with the ability to abort outstanding inner producer when the latter changes. With the addition in your snippets, the operation is all done on the system's scheduler, but the operations are not atomic, or in other words those stray cat updates could still concurrently slip in between the control signal value enqueued work & the generated event enqueued work.

> Do you perform this on the final state of the system (i.e. reducer's output)? If so, honestly, (even though it works) I see resorting to repeat skipping states on a state machine to be a bit "smelly"

For us generally it makes little difference, since:
1. our primary observation to the state machine (which renders the screen using Bento) generally tolerates repeated state;
2. in simple cases there could have no repeated state anyway; and
3. all auxiliary observations to the state machine can extract specific parts, deduplicate those and predicate over them.

> unless the state machine explicitly requires "cyclic" states (which I think most of the times is not the case?). 

Not sure what you mean by cyclic states.

> This is precisely why I was suggesting a reducer that could return an optional State

I agree from a pragmatic PoV that this should be an option, since it often makes anything around the state machine easier to write.

**Comment by andersio** [comment]
(Closed accidentally on my phone 🤦‍♂️)

**Comment by andersio** [comment]
Though even with a potentially refined implementation, it still makes state transitions decision outside the reducer & non-atomic, unless the reducer repeats the same conditional check. e.g.

```swift
static func loadNextBatchFeedback(for nearBottom: Signal<Void, NoError>) -> Feedback<State, Event> {
    return Feedback(control: nearBottom, 
                    effect: { _, state in state.canLoadNextBatch ? Event.loadNextBatch : nil })
}

static func reduce(_ state: State, _ event: Event) -> State {
    switch event {
        case .loadNextBatch where state.canLoadNextBatch:
            return .loadingNextBatch
        default:
            return state
    }
}
```

For the system to be rigid and deterministic, it is preferred to be modelled as a deterministic finite state machine, since it is a well established theory that is easy to reason & plan on paper. Under ReactiveFeedback, this generally means the reducer shall entirely represent the FSM transition table, which helps local reasoning. Another reason is that since RAF is asynchronous & concurrent (as with most similar implementations), the entire process of state transitions (essentially read-modify-write) needs to be atomic, from evaluating conditions against the current state, to actually computing the final state.

In the example above, had we not asserted `state.canLoadNextBatch` again in the reducer, we could still end up in situations where `Event.loadNextBatch` hits the reducer with a state which asserts `canLoadNextBatch == false`.

Even if the control signal is delivered on the same scheduler as the system, this would not help the non-atomic nature of the state transition. Other events can still be queued before the control signal's scheduled effect gets to be evaluated.

For example:
```
t = 0: |Effect 0|Control Signal Effect 1|
New incoming value from Control Signal, scheduled Effect 1.

t = 1: |Control Signal Effect 1|Event A|
Evaluated Effect 0, and the resulting Event A is scheduled.

t = 2: |Event A|Event B|
Evaluated Control Signal Effect 1, and the resulting Event B is scheduled.
```

So if the reducer processes Event B without rechecking all its conditions, the FSM is going to break, since Event A could have transitioned to the state by which the conditions of Event B assumed are now violated. On the other hand, if the reducer rechecks the conditions when evaluating Event B, we might as well abort the checks in Control Signal Effect 1, and defer it to the reducer.

In short, to write a system that is deterministic & rigid & maintainable with strong local reasoning, it is recommended to model after FSM and have atomic state transitions. The proposed feedback variant, however, has elements that could easily but unintentionally violate the atomicity. So from a framework author perspective, I would prefer not to include such feedback variant.

To ease the inconvenience, we may consider introducing feedback variant that bridges with external reactive sources, e.g.:

```swift
extension Feedback {
    public init<Control>(control: Signal<Control, NoError>, event: @escaping (Control) -> Event)
    public init<Control>(control: Property<Control>, event: @escaping (Control) -> Event)
}
```

**Comment by p4checo** [comment]
> With the addition in your snippets, the operation is all done on the system's scheduler, but the operations are not atomic

Can you please elaborate? What operators are truly atomic in RAS, i.e. don't leave the scheduler they're scheduled in?

> Not sure what you mean by cyclic states.

Sorry for not being clear. By "cyclic" states I mean states that have a transition to themselves. 

My point here is that if your FSM doesn't define such a state transition (A -> A), having the reducer return these transitions to then "ignore" them on the consumer side essentially "violates" the FSM definition (even though I agree that it's a pragmatic approach on some cases).

> Though even with a potentially refined implementation, it still makes state transitions decision outside the reducer & non-atomic, unless the reducer repeats the same conditional check

Repeating the same conditional check might make sense given it seems we have few guarantees about the atomicity of reactive chains for this particular use case. Are the current `Feedback` primitives themselves atomic (except the basic `init(events:)` one) ? 🤔 In all of them, the `effects` closure is used inside a `flatMap`, where the generated `SignalProducerConvertible` can be started and/or scheduled on any other scheduler before the system's scheduler, breaking the order of events too.

 > Under ReactiveFeedback, this generally means the reducer shall entirely represent the FSM transition table, which helps local reasoning.

I agree that this is the ideal solution (especially considering the asynchronous & concurrent characteristics of RAS). However, as previously stated if we always have to return a `State` from the reducer, then we should only let *valid* `Event`s be generated from the feedbacks (so that only *valid* `State` transitions are made, respecting the FSM specification). That's precisely what I am trying to accomplish here, when dealing with external sources. The current primitives seem to be more tailored to handle internal events (react to new `State`'s), and even they can not ensure full atomicity of events from what I'm seeing.

Given all these atomicity issues, I am starting to lean more and more to the solution where the reducer returns a `State?` and the system then `filterMap`s its output. The `Feedback` primitives themselves would then be "best-effort" helpers to try and only produce *valid-ish* events 😜 (because of RAS atomicity "limitations"), while the deterministic FSM behavior would be ensured by the reducer (which would now be able to discard "invalid" events).

The `Feedback`'s would keep providing all the current flexibility of generating and filtering events from internal or external sources, while preventing unnecessary work to be done (by both the reducer and the feedbacks themselves).

> In short, to write a system that is deterministic & rigid & maintainable with strong local reasoning, it is recommended to model after FSM and have atomic state transitions. The proposed feedback variant, however, has elements that could easily but unintentionally violate the atomicity. So from a framework author perspective, I would prefer not to include such feedback variant

I agree that we should model this after a FSM, and all transitions should be atomic. This proposal was an attempt to achieve that, but clearly my knowledge of RAS internals is still lacking 😅. 
Nonetheless, even though I understand if this feedback variant isn't included, I now have some doubts whether the current variants themselves aren't under similar circumstances.

> To ease the inconvenience, we may consider introducing feedback variant that bridges with external reactive sources

The variants you propose might help, but they wouldn't allow filtering values before generating the events (i.e. taking the system's current `State` into consideration). This might be the way to go if the only place where atomicity is ensured is in the reducer, though.

**Comment by andersio** [comment]
> Can you please elaborate? What operators are truly atomic in RAS, i.e. don't leave the scheduler they're scheduled in?

The effect conditionally generating the event and the reducer themselves are atomic. But them together as one entity is not mutually exclusive with any other access to the state, and therefore it is not atomic.

Even with `observe(on:)` being applied to the control signal, the said two actions are scheduled as two individuals, and other actions are free to slip in between (as long as the first-in-first-out order is maintained).

To be clear, this is not exactly an RAS issue, but a common concurrent programming problem. RAS helps in providing thread safe & atomic primitives in an push based event sourcing paradigm, hiding the need of explicit synchronisation management in most cases. But as with any atomic primitive, they can never guarantee the use of them is atomic — it is entirely up to the API consumers.

> Are the current Feedback primitives themselves atomic (except the basic init(events:) one) ?

To be fair, they are not exactly atomic. However, it is designed in a way such that when the reducer computes the new state, all previous feedback effects (derived from the now-stale state) & any outstanding event in the scheduler are discarded, and feedback effects are recreated from the new state (`flatMap(.latest)` essentially). So they are not bound to the same issue regarding state manipulation.

**Comment by p4checo** [comment]
> To be fair, they are not exactly atomic. 

I think we're getting somewhere 😄 

> However, it is designed in a way such that when the reducer computes the new state, all previous feedback effects (derived from the now-stale state) & any outstanding event in the scheduler are discarded, and feedback effects are recreated from the new state (flatMap(.latest) essentially). So they are not bound to the same issue regarding state manipulation.

So even in the current primitives, the reducer is the only place where the atomicity and deterministic FSM transitions are fully ensured (and consistency is achieved), while the rest works "with a high degree of confidence" because it relies on the "discard mechanics" provided by `state.flatMap(.latest)`, right? 🤔 

If the only way for this behaviour to be achieved is by using `state` as the "driver", it means that won't be possible with my suggested primitive (which uses the external control signal as the driver).

As an exercise, I came up with a variant of my suggestion which I think is more "atomic", but it still has the "stale state" issue because old events aren't discarded for each new `state` 😢 :

```swift
public init<Control>(control: Signal<Control, NoError>, effect: @escaping (Control, State) -> Event?) {
    self.events = { scheduler, state in
        return control.flatMap(.latest) { control in
            return state.producer
                .take(first: 1)
                .filterMap { state in effect(control, state) }
                .start(on: scheduler)
        }
    }
}
```

The produced values wouldn't be discarded for each state transition (being for each new external control instead), but the inner producer seems to be a single atomic operation on the system's scheduler (note the `start(on:)` instead of just `observe(on:)`, or both).

Starting to think it's time to give up on this new primitive (at least without an optional returning reducer 😆)
